cajita/unit_test/tstLoadBalancer.hpp

```diff
@@ -0,0 +1,519 @@
+    return ( x12 - x22 ) * ( y12 - y22 ) / 4.;
+}
+
+void lbTest3d()
+{
+    std::array<double, 8> expected_imbalance = { 0,
+                                                 0.5,
+                                                 0.66666666666666663,
+                                                 0.79999999999999993,
+                                                 0.80000000000000016,
+                                                 0.87499999999999989,
+                                                 0.8571428571428571,
+                                                 0.92857142857142871 };
+    const double imbalance_limit = 0.04;
+    const std::size_t test_steps = 2000; // Number of times to rebalance
+    const DimBlockPartitioner<3> partitioner;
+    std::array<int, 3> empty_array = { 0 };
+    std::array<int, 3> ranks =
+        partitioner.ranksPerDimension( MPI_COMM_WORLD, empty_array );
+
+    // Global mesh
+    double cell_size = 0.23;
+    std::array<int, 3> global_num_cell = { 47 * ranks[0], 38 * ranks[1],
+                                           53 * ranks[2] };
+    std::array<double, 3> global_low_corner = { 1.2, 3.3, -2.8 };
+    std::array<double, 3> global_high_corner = {
+        global_low_corner[0] + cell_size * global_num_cell[0],
+        global_low_corner[1] + cell_size * global_num_cell[1],
+        global_low_corner[2] + cell_size * global_num_cell[2] };
+    auto global_mesh = createUniformGlobalMesh(
+        global_low_corner, global_high_corner, global_num_cell );
+    std::array<bool, 3> is_dim_periodic = { false, false, false };
+
+    // Global grid
+    auto global_grid = createGlobalGrid( MPI_COMM_WORLD, global_mesh,
+                                         is_dim_periodic, partitioner );
+    int rank, size;
+    MPI_Comm_rank( MPI_COMM_WORLD, &rank );
+    MPI_Comm_size( MPI_COMM_WORLD, &size );
+    auto lb = Experimental::createLoadBalancer( MPI_COMM_WORLD, global_grid );
+    std::array<double, 6> vertices = lb->getVertices();
+    std::array<double, 6> internal_vertices = lb->getInternalVertices();
+    double work = lbWork( vertices );
+    global_grid =
+        lb->createBalancedGlobalGrid( global_mesh, partitioner, work );
+    double imbalance = lb->getImbalance();
+    if ( rank == 0 )
+        printf( "LB Imbalance: %.17g\n", imbalance );
+    if ( size < static_cast<int>( expected_imbalance.size() ) )
+    {
+        EXPECT_DOUBLE_EQ( imbalance, expected_imbalance[size - 1] );
+    }
+    for ( std::size_t step = 1; step < test_steps; ++step )
+    {
+        work = lbWork( vertices );
+        global_grid =
+            lb->createBalancedGlobalGrid( global_mesh, partitioner, work );
+        vertices = lb->getVertices();
+        internal_vertices = lb->getInternalVertices();
+        imbalance = lb->getImbalance();
+        // if ( rank == 0 )
+        //    printf( "LB Imbalance: %g\n", imbalance );
+        if ( imbalance < imbalance_limit )
```

Comment:
Do I understand correctly that large numbers of ranks would require more steps to balance? @sslattery what do you think about a while loop until it hits the threshold vs a fixed maximum number of test steps the will fail for the largest cases?

Reply:
Exactly, without a minimum domain width the threshold will be reached, but it may take more steps. If a minimum domain width is set the same should be true in our setup as the number of cells per domain is constant and hence the loadbalancing resolution as well.

The `test_steps` limit is just there to avoid a possible infinite loop. But we can increase the limit, since it only affects large numbers of ranks anyway.